### ISSUE TITLE
X509_STORE_CTX_purpose_inherit semantics details

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#18801]
Three questions:
* Should `X509_STORE_CTX_purpose_inherit` crash on `ctx` or `ctx->param` being NULL?
* Should it really preserve `ctx->param->purpose` if this already contains a defined purpose value?
* Should it really preserve `ctx->param->trust` if this already contains a defined trust value?

**Comment by kaduk** [comment]
> Three questions:
> 
> * Should `X509_STORE_CTX_purpose_inherit` crash on `ctx` or `ctx->param` being NULL?

This seems to be just another incarnation of our recurring discussion about explicit NULL checks, with (IIRC) the balance being that we don't plan to provide comprehensive checks against application bugs.

> * Should it really preserve `ctx->param->purpose` if this already contains a defined purpose value?
> * Should it really preserve `ctx->param->trust` if this already contains a defined trust value?

There's a conceptual sense for what this API is trying to do in which this behavior makes sense (something roughly like "before we actually use this object, pull in the default values for anything not set explicitly").  But I think the question as written is not actually worth answering -- these APIs have been around and undocumented for so long that the only option for responsible maintainership is to retain their functionality unchanged until we remove them entirely.  If we do have a chance to think about what an actually good API would look like and what we want from it, we should add a new API for that behavior rather than trying to change the existing one.

(IMHO this API feels like something that would only be used internally to the library, and when it was written there was no distinction between public and private functions.  But we are constrained by our stability policy.)

**Comment by mattcaswell** [comment]
> Should X509_STORE_CTX_purpose_inherit crash on ctx or ctx->param being NULL?
>
> This seems to be just another incarnation of our recurring discussion about explicit NULL checks, with (IIRC) the balance being that we don't plan to provide comprehensive checks against application bugs.

I agree for `ctx`. I'm not sure that I agree for `ctx->param`. It seems sensible to me to verify that this is non-NULL before use.

**Comment by DDvO** [comment]
Reopening this as a reminder on Matt's above comment.

**Comment by DDvO** [comment]
Re-opening again because I suppose the following according to Matt's above comment is not yet done:

> for `ctx->param`. It seems sensible to me to verify that this is non-NULL before use.

**Comment by t8m** [comment]
ctx->param should be always non-NULL, so no need to check it. And please note it is nowhere checked.
